### PR TITLE
KAFKA-10766: Unit test cases for RocksDBRangeIterator

### PR DESCRIPTION
**Pull request by vamossagar12**
This PR aims to add unit test cases for RocksDBRangeIterator which were missing.

### CODE REVIEW

**Comment by vamossagar12** [comment]
@cadonna , this is post our discussion on [prefix-scan-pr](https://github.com/apache/kafka/pull/9508). Plz review whenever you find the time.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java]: It is not a good idea to change a class due to unit tests if it is not absolutely necessary. In this case it is definitely not necessary. See my suggestion how to change the tests.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: You usually do not want to mock the class under test, because you want to test it. Also partial mocks should only be used if absolutely necessary. A rule of thumb is if a partial mock is needed then most probably the design has a flaw.
In this specific case, you should mock RocksDB's iterator. For the class under test, you should test `hasNext()`, `next()`, `peekNextKey()` and `close()`, because those are the one exposed (`makeNext()` should actually de declared as `protected`, IMO). 

```suggestion
        final String key1 = "a";
        final String key2 = "b";
        final String key3 = "c";
        final String key4 = "d";
        final String value = "value";

        final  Bytes key1Bytes = Bytes.wrap(key1.getBytes());
        final  Bytes key2Bytes = Bytes.wrap(key2.getBytes());
        final  Bytes key3Bytes = Bytes.wrap(key3.getBytes());
        final  Bytes key4Bytes = Bytes.wrap(key4.getBytes());
        final byte[] valueBytes = value.getBytes();

        final RocksIterator rocksIterator = mock(RocksIterator.class);
        rocksIterator.seek(key1Bytes.get());
        expect(rocksIterator.isValid())
            .andReturn(true)
            .andReturn(true)
            .andReturn(true)
            .andReturn(true)
            .andReturn(false);
        expect(rocksIterator.key())
            .andReturn(key1Bytes.get())
            .andReturn(key2Bytes.get())
            .andReturn(key3Bytes.get())
            .andReturn(key4Bytes.get());
        expect(rocksIterator.value()).andReturn(valueBytes).times(4);
        rocksIterator.next();
        expectLastCall().times(4);
        replay(rocksIterator);

        final RocksDBRangeIterator rocksDBRangeIterator = new RocksDBRangeIterator(
            STORE_NAME,
            rocksIterator,
            Collections.emptySet(),
            key1Bytes,
            key4Bytes,
            true
        );
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key1Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key2Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key3Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key4Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(false));
        verify(rocksIterator);
```

**Review by cadonna — COMMENTED** [review]
@vamossagar12 Thank you for the PR!

Here my feedback!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: @cadonna , honestly even I wasn't making the changes that I ended up making but I assumed that makeNext is the only method that needs to be tested wrt Range Iterator as other methods are inherited by it so those should have test cases separately. I have changes the test cases now.

Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: @cadonna , did you get  a chance to review these changes?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]:
```suggestion
        final RocksDBRangeIterator rocksDBRangeIterator = new RocksDBRangeIterator(
            storeName,
            rocksIterator,
            Collections.emptySet(),
            key1Bytes,
            key3Bytes,
            true
        );
```
Please also fix the other wrong indentations.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]:
```suggestion
        final Bytes toBytes = Bytes.wrap(to.getBytes());
```
Please also check other occurrences of this additional space.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: nit: Couldn't you Instead of using a literal, use `Bytes.increment(key4Bytes)` for the `to` parameter?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]:
```suggestion
        expect(rocksIterator.isValid()).andReturn(false);
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java:343]: I think, it would be good to verify that a second call to  `peekNextKey()` right after the first call to `peekNextKey()` returns the same value, since this is the main difference between `next()` and `peekNextKey()`.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java:381]: See my comment above.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: Please remove empty lines here and in the other test methods.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: Please rename to `shouldCloseIterator()`.

**Review by cadonna — COMMENTED** [review]
@vamossagar12 Thank you for the updates!

Here my feedback!

I have mainly minor comments and one major about testing `peekNextKey()`.

**Comment by cadonna** [comment]
> honestly even I wasn't making the changes that I ended up making but I assumed that makeNext is the only method that needs to be tested wrt Range Iterator as other methods are inherited by it so those should have test cases separately.

Fair point. I would agree, if we had tests for `RocksDbIterator`, but we haven't. I thought, it would be easier to restrict the tests to the range iterator and in a second step refactor the code to test also the parent class.

**Comment by vamossagar12** [comment]
@cadonna , done. Also, I need to create a new ticket to add the tests for prefix scan related changes. I will add them once the KIP-614 is merged.

**Comment by vamossagar12** [comment]
@cadonna , i have added the test case for prefix scan as well.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: If you change the `toInclusive` to `true` in this test, the test also passes. That means, that you are not testing `toInclusive` in this test.

I suggest the following test: 
```
    public void shouldExcludeEndOfRange() {
        final RocksIterator rocksIterator = mock(RocksIterator.class);
        rocksIterator.seek(key1Bytes.get());
        expect(rocksIterator.isValid())
            .andReturn(true)
            .andReturn(true);
        expect(rocksIterator.key())
            .andReturn(key1Bytes.get())
            .andReturn(key2Bytes.get());
        expect(rocksIterator.value()).andReturn(valueBytes).times(2);
        rocksIterator.next();
        expectLastCall().times(2);
        replay(rocksIterator);
        final RocksDBRangeIterator rocksDBRangeIterator = new RocksDBRangeIterator(
            storeName,
            rocksIterator,
            Collections.emptySet(),
            key1Bytes,
            key2Bytes,
            true,
            false
        );
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key1Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(false));
        verify(rocksIterator);
    }
```

**Review by cadonna — COMMENTED** [review]
@vamossagar12 Thanks!

I have just one major comment.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java
Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java]: @cadonna , added this test.

**Review by cadonna — APPROVED** [review]
Thanks, @vamossagar12 !
LGTM!

**Comment by cadonna** [comment]
Call for committer approval: @guozhangwang @ableegoldman @vvcephei @mjsax @abbccdda

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk and cherry-picked to 2.8 cc @vvcephei

**Pull request merged**
